pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/MaximumPinotConsumptionDelayTracker.java

```diff
@@ -0,0 +1,306 @@
+ * Highlights:
+ * 1-An object of this class is hosted by each RealtimeTableDataManager.
+ * 2-The object tracks ingestion delays for all partitions hosted by the current server for the given Realtime table.
+ * 3-Partition delays are updated by all LLRealtimeSegmentDataManager objects hosted in the corresponding
+ *   RealtimeTableDataManager.
+ * 4-The class tracks the maximum of all ingestion delays observed for all partitions of the given table.
+ * 5-A Metric is derived from reading the maximum tracked by this class.
+ * 6-Delay reported for partitions that do not have events to consume is reported as zero.
+ * 7-We track the time at which each delay sample was collected so that delay can be increased when partition stops
+ *   consuming for any reason other than no events being available for consumption.
+ * 8-Segments that go from CONSUMING to DROPPED states stop being tracked so their delays do not cloud
+ *   delays of active partitions.
+ * 9-When a segment goes from CONSUMING to ONLINE, we start ticking time for the corresponding partition.
+ *   If no consumption is noticed after a timeout, then we read ideal state to confirm the server still hosts the
+ *   partition. If not, we stop tracking the respective partition.
+ * 10-A timer thread is started by this object to track timeouts of partitions and drive the reading of their ideal
+ *    state.
+ *
+ *  The following diagram illustrates the object interactions with main external APIs
+ *
+ *     (CONSUMING -> ONLINE state change)
+ *             |
+ *      markPartition(partitionId)      |<-recordPinotConsumptionDelay()-{LLRealtimeSegmentDataManager(Partition 0}}
+ *            |                         |
+ * ___________V_________________________V_
+ * |           (Table X)                |<-recordPinotConsumptionDelay()-{LLRealtimeSegmentDataManager(Partition 1}}
+ * | MaximumPinotConsumptionDelayTracker|           ...
+ * |____________________________________|<-recordPinotConsumptionDelay()-{LLRealtimeSegmentDataManager (Partition n}}
+ *              ^                      ^
+ *              |                       \
+ *   tickInactivePartitions()    stopTrackingPinotConsumptionDelay(partitionId)
+ *    _________|__________                \
+ *   | TimerTrackingTask |          (CONSUMING -> DROPPED state change)
+ *   |___________________|
+ *
+ */
+
+public class MaximumPinotConsumptionDelayTracker {
+  /*
+   * Class to handle timer thread that will track inactive partitions
+   */
+  private class TrackingTimerTask extends TimerTask {
+    private String _name;
+    private MaximumPinotConsumptionDelayTracker _tracker;
+
+    public TrackingTimerTask(String name, MaximumPinotConsumptionDelayTracker tracker) {
+      _name = name;
+      _tracker = tracker;
+    }
+
+    @Override
+    public void run() {
+      // tick inactive partitions every interval of time to keep tracked partitions up to date
+      _tracker.tickInactivePartitions();
```

Comment:
It looks like isPartitionHostedByThisServerPerIdealState can potentially throw an exception (e.g. due to an underlying Helix error like a connection error), which would kill this TimerTask's thread. To avoid that, I suggest wrapping with a try/catch here.

Reply:
done.